### PR TITLE
Adapt transform_inclusive_scan to C++ 20

### PR DESCRIPTION
**Pull request by Jedi18**
Adapted transform_inclusive_scan to C++ 20 and added container overloads. (range and sentinel). Added container tests. Seperated segmented overloads.

## Any background context you want to provide?
Issue #4822
Issue #3646
Issue #1668
Issue #5156

### CODE REVIEW

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/transform_inclusive_scan.hpp [hunk 1]
```diff
@@ -186,330 +578,249 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+            ExPolicy>::value&& hpx::traits::is_iterator<FwdIter1>::value&&
+                hpx::traits::is_iterator<FwdIter2>::value&& hpx::is_invocable_v<
+                    Conv, typename std::iterator_traits<FwdIter1>::value_type>&&
+                    hpx::is_invocable_v<Op,
+                        typename hpx::util::invoke_result<Conv,
+                            typename std::iterator_traits<
+                                FwdIter1>::value_type>::type,
+                        typename hpx::util::invoke_result<Conv,
+                            typename std::iterator_traits<
+                                FwdIter1>::value_type>::type>)>
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/transform_inclusive_scan.hpp:618]: Would you mind making this a bit more readable?

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/transform_inclusive_scan.hpp [hunk 2]
```diff
@@ -186,330 +578,249 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/transform_inclusive_scan.hpp:657]: Same here, please apply some more decent formatting.

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/transform_inclusive_scan.hpp [hunk 3]
```diff
@@ -186,330 +578,249 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+        // clang-format off
+        template <typename InIter, typename OutIter, typename BinOp,
+            typename UnOp,
+            HPX_CONCEPT_REQUIRES_(
+                hpx::traits::is_iterator<InIter>::value &&
+                hpx::traits::is_iterator<OutIter>::value &&
+                hpx::is_invocable_v<UnOp,
+                    typename std::iterator_traits<InIter>::value_type> &&
+                hpx::is_invocable_v<BinOp,
+                    typename hpx::util::invoke_result<UnOp,
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/transform_inclusive_scan.hpp:699]: FWIW, we do have `hpx::invoke_result_t<>` (https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/functional/include/hpx/functional/invoke_result.hpp#L41-L42)

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/transform_inclusive_scan.hpp [hunk 4]
```diff
@@ -0,0 +1,1069 @@
...
+                    typename std::iterator_traits<InIter>::value_type> &&
+                hpx::is_invocable_v<BinOp,
+                    typename hpx::util::invoke_result<UnOp,
+                        typename std::iterator_traits<InIter>::value_type>::type,
+                    typename hpx::util::invoke_result<UnOp,
+                        typename std::iterator_traits<InIter>::value_type>::type
+                >
+            )>
+        // clang-format on
+        friend OutIter tag_fallback_dispatch(
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/transform_inclusive_scan.hpp:818]: Same comment as for `inclusive_scan` and `exclusive_scan`: should we return `in_out_iterator<InIter, OutIter>` to refer to the sentinel's position on return?

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/transform_inclusive_scan.hpp [hunk 5]
```diff
@@ -0,0 +1,1069 @@
...
+                        hpx::traits::is_iterator<
+                            OutIter>::value&& hpx::is_invocable_v<UnOp,
+                            typename std::iterator_traits<InIter>::value_type>&&
+                            hpx::is_invocable_v<BinOp,
+                                typename hpx::util::invoke_result<UnOp,
+                                    typename std::iterator_traits<
+                                        InIter>::value_type>::type,
+                                typename hpx::util::invoke_result<UnOp,
+                                    typename std::iterator_traits<
+                                        InIter>::value_type>::type>)>
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/transform_inclusive_scan.hpp:948]: formatting?

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/transform_inclusive_scan.hpp [hunk 6]
```diff
@@ -86,53 +484,57 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
-                typename Conv, typename T, typename Op>
-            static OutIter sequential(ExPolicy, InIter first, InIter last,
-                OutIter dest, Conv&& conv, T&& init, Op&& op)
+            template <typename ExPolicy, typename InIter, typename Sent,
+                typename OutIter, typename Conv, typename T, typename Op>
+            static constexpr util::in_out_result<InIter, OutIter> sequential(
+                ExPolicy, InIter first, Sent last, OutIter dest, Conv&& conv,
+                T&& init, Op&& op)
             {
                 return sequential_transform_inclusive_scan(first, last, dest,
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/transform_inclusive_scan.hpp:503]: I'm surprised this compiles. `sequential_transform_inclusive_scan` returns a single iterator that shouldn't be convertible to an `in_out_result`

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/transform_inclusive_scan.hpp [hunk 7]
```diff
@@ -86,53 +484,57 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
-                typename Conv, typename Op>
-            static OutIter sequential(ExPolicy&&, InIter first, InIter last,
-                OutIter dest, Conv&& conv, Op&& op)
+            template <typename ExPolicy, typename InIter, typename Sent,
+                typename OutIter, typename Conv, typename Op>
+            static constexpr util::in_out_result<InIter, OutIter> sequential(
+                ExPolicy&&, InIter first, Sent last, OutIter dest, Conv&& conv,
+                Op&& op)
             {
                 return sequential_transform_inclusive_scan_noinit(first, last,
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/transform_inclusive_scan.hpp:514]: Same here, I'm surprised this compiles. Do we have a test for this?

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/transform_inclusive_scan.hpp (hunk 6 above)
Comment (hkaiser) [reply libs/parallelism/algorithms/include/hpx/parallel/algorithms/transform_inclusive_scan.hpp:503]: I take this back, sorry for the noise.

**Review by hkaiser — DISMISSED** [review]
LGTM, thanks!

**Comment by Jedi18** [comment]
Merge order - exlcusive_scan -> transform_exclusive_scan -> tranform_inclusive_scan

**Pull request merged**